draft/mson-parameters-headers.md

```diff
@@ -0,0 +1,216 @@
+---
+RFC: XXXX
+Author: Z
+Status: Draft
+Created: 2015-09-23
+Last Modified: 2015-09-28
+---
+
+# API Blueprint RFC XXXX: MSON Parameters and Headers
+
+## Table of Contents
+
+- [Abstract](#abstract)
+- [Motivation](#motivation)
+- [Implementation](#implementation)
+- [Backwards Compatibility](#backwards-compatibility)
+
+## Abstract
+
+This RFC proposes using full [MSON][] syntax for description of URI parameters
+and HTTP headers.
+
+## Motivation
+
+Using MSON syntax for description of URI parameters and HTTP Headers will
+greatly help to unify the API Blueprint syntax and introduce reusability of
+URI parameters and HTTP headers. It will also greatly introduce the reusability
+of other data structures.
+
+In addition it will address many issues previously tracked on API Blueprint
+GitHub repository.
+
+### URI Parameters Issues Tracked
+
+This RFC addresses following issues:
+
+1. [Usage of data structures in Parameters](https://github.com/apiaryio/api-blueprint/issues/250)
+1. [Global query URI parameter](https://github.com/apiaryio/api-blueprint/issues/24)
+1. [Description of enumeration elements](https://github.com/apiaryio/api-blueprint/issues/119)
+1. [Array in parameters](https://github.com/apiaryio/api-blueprint/issues/240)
+1. [Minimum/Maximum values for URI Parameters](https://github.com/apiaryio/api-blueprint/issues/211)
+
+    NOTE: This is going to be addressed once MSON introduces validation.
+
+### HTTP Headers Issues Tracked
+
+This RFC addresses following issues:
+
+1. [Parametric Headers Description](https://github.com/apiaryio/api-blueprint/issues/26)
+1. [Referenced HTTP headers](https://github.com/apiaryio/api-blueprint/issues/17)
+1. [Support for multiline syntax header](https://github.com/apiaryio/api-blueprint/issues/60)
```

Comment:
We would need to think about multiple headers with the same key (https://github.com/apiaryio/api-blueprint/issues/345). Especially with how that will work with inheritance and mixins.

Example:

```apib
## Chapter [/TheBook/chapter{id}]
### View a chapter [GET]

+ Response 204
    + Headers
        + Include Pagination
        + Include Creatable

## Data Structures
### Pagination
+ Link (array)
    + `<http://example.com/TheBook/chapter1>; rel="prev"; title="previous chapter"`
    + `<http://example.com/TheBook/chapter3>; rel="next"; title="next chapter"`

### Creatable
+ Link: `<http://example.com/TheBook>; rel="create"; title="create a new chapter"`
```

The user would probably expect all three link headers to be present in the response, the pagination mixing and creatable mixin. Normally mixins in MSON would override the key names and only `create` link header would be present.